### PR TITLE
Launch Template Support

### LINKED ISSUE #17: EPIC: support launch templates
instance-manager should support launch templates.
This can be done as a separate kind, e.g. `Kind: InstanceGroupTemplate` (or different), and can then reference it in InstanceGroup e.g. `.spec.provisioner.configuration.template`.

### LINKED ISSUE #164: Launch Templates - AWS API Calls & Caching
Ref #17 

In order to support launch templates, we need the following wrapper methods in AwsWorker

- CreateLaunchTemplate (CREATE)
- DeleteLaunchTemplate (DELETE)
- CreateTemplateVersion (UPDATE)

- DescribeAutoscalingLaunchTemplates (READ)
- DescribeTemplateVersion (READ)

For the Describe calls we can define caching of 60 seconds (same as launch config).

### PR DESCRIPTION
**Pull request by eytan-avisror**
Closes #17 
Closes #164 

- [x] Implementation/refactor
- [x] Add unit tests
- [x] Add functional test
- [x] Manual acceptance/stress/load testing
- [x] Documentation
- [x] Review / Cleanup

### CODE REVIEW

**Comment by preflightsiren** [comment]
@eytan-avisror having Launch Template support is the last blocker for my team adopting instance-manager and moving to EKS; I have some time to dedicate on developing this feature, but obviously I can see that it's already started.

Is there any work I can take on to help with the development on this feature? eg. https://github.com/keikoproj/instance-manager/issues/164

**Comment by eytan-avisror** [comment]
Thanks @preflightsiren !
Unfortunately this is one huge PR so not sure it will be easy to collaborate on this. This work is almost done - still have some bugs to fix and tests to write.. hope to get this merged in the next 2 weeks 👍

**Comment by preflightsiren** [comment]
Sounds good. If you keep pushing your latest to GitHub, I'll keep pulling it and running it against our clusters to test it out 😄

File: controllers/provisioners/eks/scaling/launchtemplate_test.go
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/launchtemplate_test.go:603]: Is there a reason you don't use `testing.Run()` ?

I was thinking maybe adding a `name: string` or `description: string` to the tests struct that can describe the test case.

then the output becomes:
```
TestLaunchTemplateDrifted/volume-size-should-drift    ok
```

File: controllers/provisioners/eks/scaling/launchtemplate_test.go
Comment (preflightsiren) [inline controllers/provisioners/eks/scaling/launchtemplate_test.go:573]: why do you always have to set the `IamInstanceProfile`?

Comment (eytan-avisror) [reply controllers/provisioners/eks/scaling/launchtemplate_test.go:573]: Test would get a NPE (in launchtemplate.go:204) otherwise because AWS structs are all pointers.
In this case it's OK to assume it will never be nil, because an IAM role is always added to the launchtemplate.

I will do more cleanup before merging to remove repeating code.

Comment (preflightsiren) [reply controllers/provisioners/eks/scaling/launchtemplate_test.go:573]: I've got some fixes for those in https://github.com/eytan-avisror/instance-manager/pull/3

**Comment by eytan-avisror** [comment]
This is ready for review now

**Comment by eytan-avisror** [comment]
@preflightsiren @backjo 
feel free to review code and provide suggestions if you'd like

File: controllers/provisioners/eks/eks_test.go
Comment (preflightsiren) [inline controllers/provisioners/eks/eks_test.go:132]: do we ever test the default case of `Type` being undefined?

File: docs/EKS.md
Comment (preflightsiren) [inline docs/EKS.md:116]: 👍

**Review by preflightsiren — APPROVED** [review]
LGTM.

File: controllers/providers/aws/aws.go
Comment (eytan-avisror) [reply controllers/providers/aws/aws.go:76]: I noticed a bug where there is a race condition if IAM doesn't fully propagate nodes will not come up.
This is a real pain, because there is no way around this other than blindly waiting, since the IAM role is 'created' but it still needs some time to propagate before it's used to create a scaling group.
IAM APIs are super old, and some aspects of it are really annoying.

This extra 10 seconds seems to resolve the issue.
This is only an issue when you don't provide your own IAM role and expect the controller to create it

File: controllers/common/utils.go
Comment (eytan-avisror) [reply controllers/common/utils.go:238]: Good catch, i'll address it

File: controllers/provisioners/eks/cloud.go
Comment (eytan-avisror) [reply controllers/provisioners/eks/cloud.go:61]: Not sure which line you're referring to - mixedInstances is not a member of `EksInstanceGroupContext`, can you take another look?

mixedInstances in the location you annotated is a local variable that is set to `configuration.GetMixedInstancesPolicy()`.

I will change the variable name to mixedInstancesPolicy - more accurate

File: controllers/provisioners/eks/eks_test.go
Comment (eytan-avisror) [reply controllers/provisioners/eks/eks_test.go:132]: This happens in instancegroup_types.go:360, we SHOULD add unit tests around all of the validations we do, but since this is a larger scope and this is manually tested, I will create an issue to address this

Comment (eytan-avisror) [reply controllers/provisioners/eks/eks_test.go:132]: #192 opened

File: controllers/provisioners/eks/cloud.go
Comment (preflightsiren) [reply controllers/provisioners/eks/cloud.go]: ahh, my mistake. going blind :P

Comment (eytan-avisror) [reply controllers/provisioners/eks/cloud.go]: 😆

**Comment by preflightsiren** [comment]
What's left to getting this PR ready to merge?

**Comment by eytan-avisror** [comment]
> What's left to getting this PR ready to merge?

Waiting for reviews from authorized-approvers, since this is a big PR it might some time.
If no one reviews/approves it in a few days I will force-merge 💪

File: controllers/providers/kubernetes/crd.go
Comment (kevdowney) [inline controllers/providers/kubernetes/crd.go:115]: Debug statement?

Comment (eytan-avisror) [reply controllers/providers/kubernetes/crd.go:115]: yep, will remove 😉

**Pull request merged**
